### PR TITLE
libs: Add envoy.dependency.cve_scan

### PR DESCRIPTION
**Pull request by phlax**

### CODE REVIEW

**Comment by phlax** [comment]
cc @htuch @moderation

this PR is still WIP, but the code is mostly there, and type hinting is added, so mostly just writing some unit tests and any refactoring that goes with that

it includes filtering on CVE versions were possible

**Comment by htuch** [comment]
Looks great, thanks for taking on this major cleanup. I think one thing I'd suggest is beyond unit tests, having a few "integration" tests that use real CVE database content, since historically, a lot of the issues we've hit are places where the CVE database has changed the fields it uses or some corner case, and we'd want to capture future regressions like this.

**Comment by phlax** [comment]
> ...having a few "integration" tests that use real CVE database content...

yep, sure ill add once i have completed the unit tests

i guess it would be helpful to do something with the repository_locations data also

File: envoy.dependency.cve_scan/tests/test_integration.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
+
```
Comment (phlax) [inline envoy.dependency.cve_scan/tests/test_integration.py:1]: @htuch ive added integration tests here

it adds quite a lot of code for the test data, but otoh it provides a reference implementation which is quite helpful

working with the nist data is a pita as its huge and unwieldy, but ill fish some more CVEs out of it and add some further tests

there are probably some cleanups also, but would like to "move to close" on this PR

File: envoy.dependency.cve_scan/tests/test_integration.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        return DummyCVE
+
+    @property
+    def dependency_class(self):
+        return DummyDependency
+
+    # implementations should cache this
+    @property
+    def dependency_metadata(self):
+        # this needs to be mocked for these tests
```
Comment (phlax) [inline envoy.dependency.cve_scan/tests/test_integration.py:73]: s/this/super/

File: envoy.dependency.cve_scan/tests/test_integration.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    f"nist:{p.name.split('.')[0]}":
+    p for p in list(nist.glob("*.json"))}
+deps = pathlib.Path("envoy.dependency.cve_scan/tests/integration/deps")
+DEPS_TEST_DATA = {
+    f"deps:{p.name.split('.')[0]}": p
+    for p in list(deps.glob("*.json"))}
+outcomes = pathlib.Path("envoy.dependency.cve_scan/tests/integration/outcomes")
+OUTCOME_TEST_DATA = {
+    f"outcome:{p.name.split('.')[0]}": p
+    for p in list(outcomes.glob("*.json"))}
```
Comment (htuch) [inline envoy.dependency.cve_scan/tests/test_integration.py:105]: Is there a way to refactor this to avoid boilerplate?

File: envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/checker.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+import aiohttp
+
+import abstracts
+
+from aio.functional import async_property
+from aio.tasks import concurrent
+
+from envoy.base import checker, utils
+
+from ..exceptions import CVECheckError
```
[the remark this answers is not in the document]
Comment (phlax) [reply envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/checker.py:21]: perhaps in the case of `..` as it makes modules less portable - but im not so sure about sibling imports - i think that `.` is the best mostly for the same reason - ie it makes the modules more portable

ill fix the `..` and maybe make some of the imports more like the style suggestion in the google style doc...

File: envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/checker.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+from .cve import ACVE
+from .dependency import ADependency
+
+
+NIST_URL_TPL = (
+    "https://nvd.nist.gov/feeds/json/cve/1.1/nvdcve-1.1-{year}.json.gz")
+SCAN_FROM_YEAR = 2018
+
+
+class ACVEChecker(checker.AsyncChecker, metaclass=abstracts.Abstraction):
```
[the remark this answers is not in the document]
Comment (phlax) [reply envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/checker.py:33]: yep, agreed - ill dress the abstract classes with comments now...

File: envoy.dependency.cve_scan/tests/test_integration.py (hunk 3 above)
Comment (phlax) [reply envoy.dependency.cve_scan/tests/test_integration.py:105]: iiuc the suggestion probably yes - i can just make them all into a single `TEST_DATA` dictionary and populate dynamically...

File: envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/cve.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        """Determine whether a CPE matches according to installed version of a
+        dependency."""
+        return (
+            str(cpe) in self.tracked_cpes
+            and (
+                self.version_matcher_class(cpe_match)(
+                    self.tracked_cpes[str(cpe)])))
+
+    def parse_cve_date(self, date_str: str) -> date:
+        if not date_str.endswith('Z'):
```
Comment (phlax) [inline envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/cve.py:173]: @htuch this was one of my concerns

i have switched the `assert` -> `Exception` 

im still thinking that this is not required unless someone is using this to read some non-NIST data that is in NIST format, but for some reason doesnt use UTC times (seems unlikely to me)

File: envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/cpe.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            # An exact version match is a hit.
+            or self.version == dep.version
+            # Allow the 'release_date' dependency metadata to substitute
+            # for date.
+            # TODO(htuch): Consider fuzzier date ranges.
+            or self.version == dep.release_date)
+        if version_match:
+            return True
+
+        # Try a fuzzy date match to deal with versions like fips-20190304 in
```
Comment (phlax) [inline envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/cpe.py:82]: @htuch this is my other concern

not sure how this code was hit in the past but afaict in the current implementation (that this is replacing) it never gets hit

i believe this is because none of the cpes listed in repository_locations.bzl are versioned - cf:

```console
$ git grep cpe bazel/repository_locations.bzl | grep -v "N/A"
```

Im wondering whether we want to keep given its not used

File: envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/cpe.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        # useful for
+        # LLVM where multiple product need to be covered.
+        matching_parts = (
+            self.part == dep_cpe.part
+            and self.vendor == dep_cpe.vendor
+            and (
+                dep_cpe.product == '*'
+                or self.product == dep_cpe.product))
+        if not matching_parts:
+            return False
```
Comment (phlax) [inline envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/cpe.py:67]: afaict this line is also never hit

File: envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/cpe.py (hunk 7 above)
Comment (htuch) [reply envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/cpe.py:82]: It wasn't the repository_locationz.bzl CPE that was versioned, rather the version field in something like https://github.com/envoyproxy/envoy/blob/aac4631c26745de1a39556abc68bd1d3ed95aa2f/bazel/repository_locations.bzl#L97 was not a real "version" which could be used to compare against the CPE version in the NIST database. Hence this.

File: envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/cve.py (hunk 6 above)
Comment (htuch) [reply envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/cve.py:173]: Yeah, I think assuming NIST actual use for format is fine. If you want to be defensive, leave the assert.

**Review by htuch — COMMENTED** [review]
LGTM, ship it! I feel this code, for better-or-worse, is the best we have for CVE scans for now, and it's in a much more maintainable state.

File: envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/cpe.py (hunk 7 above)
Comment (phlax) [reply envoy.dependency.cve_scan/envoy/dependency/cve_scan/abstract/cpe.py:82]: ok, ill leave it there for now, its not tested, but we can revisit - i want to add some more tests after anyway

**Comment by phlax** [comment]
> ... ship it! I feel this code, for better-or-worse, is the best we have...

yep, agreed - lets bed it in on the envoy side, tests are ~99% - the bit that isnt tested is that bit i highlighted above, so easy to track
